app/src/main/kotlin/com/simplemobiletools/gallery/activities/ViewPagerActivity.kt

```diff
@@ -454,52 +455,60 @@ class ViewPagerActivity : SimpleActivity(), ViewPager.OnPageChangeListener, View
+                try {
+                    val bitmap = BitmapFactory.decodeFile(currPath)
+                    getFileOutputStream(tmpFile) {
+                        if (it == null) {
+                            toast(R.string.unknown_error_occurred)
+                            deleteFile(tmpFile) {}
+                            return@getFileOutputStream
                         }
-                    } else {
-                        copyFile(currPath, it)
-                        rotateFileByExif(it)
+
+                        if (currPath.isJpg()) {
+                            saveRotation(getCurrentFile(), tmpFile)
+                        } else {
+                            saveFile(tmpFile, bitmap, it as FileOutputStream)
+                        }
+
+                        if (tmpFile.length() > 0 && selectedFile.exists()) {
+                            deleteFile(selectedFile) {}
+                        }
+                        copyFile(tmpFile, selectedFile)
+                        scanPath(selectedFile.absolutePath) {}
+                        deleteFile(tmpFile) {}  // Not necessary?
+                        toast(R.string.file_saved)
+
+                        it.flush()
+                        it.close()
+                        mRotationDegrees = 0f
+                        invalidateOptionsMenu()
                     }
-                } else {
-                    rotateFileByDegrees(currPath, it)
+                } catch (e: OutOfMemoryError) {
+                    toast(R.string.out_of_memory_error)
+                    deleteFile(tmpFile) {}
+                } catch (e: Exception) {
+                    showErrorToast(e)
+                    deleteFile(tmpFile) {}
                 }
             }).start()
         }
     }
 
-    private fun rotateFileByDegrees(sourcePath: String, destinationPath: String) {
-        val tmpFile = File(File(destinationPath).parent, ".tmp_${destinationPath.getFilenameFromPath()}")
+    private fun copyFile(source: File, destination: File) {
+        var inputStream: InputStream? = null
+        var out: OutputStream? = null
         try {
-            getFileOutputStream(tmpFile) {
-                if (it == null) {
-                    toast(R.string.unknown_error_occurred)
-                    return@getFileOutputStream
-                }
-
-                val bitmap = BitmapFactory.decodeFile(sourcePath)
-                saveFile(tmpFile, bitmap, it as FileOutputStream)
-                it.flush()
-                it.close()
-
-                val destination = File(destinationPath)
-                deleteFile(destination) {
-                    renameFile(tmpFile, destination) {}
-                }
-
-                toast(R.string.file_saved)
-                mRotationDegrees = 0f
-                invalidateOptionsMenu()
+            if (needsStupidWritePermissions(destination.absolutePath)) {
+                getFileDocument(destination.parent)
             }
```

Comment:
this getFileDocument block isnt needed, its not doing anything, just slowing down the copy by getting a useles filedocument

Reply:
I'm not sure, but maybe ```if (needsStupidWritePermissions(destination.absolutePath)) {
  getFileDocument(destination.parent)
}``` is actually still necessary! I just got the following:
- Installed Gallery anew
- Installed a new SD card
- Tried to rotate an save to SD card. I got a toast message "invalid URI".
- It turned out that Simple Gallery didn't have the SD card write permissions yet, _and didn't ask for it either_.
- In another operation (I believe deleting a photo), I _did_ get the permission dialog, had to go to Android's File Manager and select the SD card.
- Afterwards, the exact same rotate operation that previously failed, worked flawlessly.

Reply:
it wouldnt be solved properly by this block. Adding a handleSAFDialog(selectedFile) {...} after creating "selectedFile" and wrapping everything after it in the new brackets should do it.